File: ios-localwave/LocalWave/AppState/SettingsViewModel.swift
import Foundation

@MainActor
public final class SettingsViewModel: ObservableObject {
    @Published public var displayName: String
    @Published public var channelText: String
    @Published public private(set) var identity: LocalIdentity?
    @Published public private(set) var transportState = TransportState()
    @Published public private(set) var permissionSnapshot = PermissionSnapshot()
    @Published public private(set) var nearbyPeerCount = 0
    @Published public private(set) var messageStatusCounts: [MessageStatus: Int] = [:]
    @Published public private(set) var errorMessage: String?

    public let store: AppStore
    private var peerTask: Task<Void, Never>?
    private var transportTask: Task<Void, Never>?

    public init(store: AppStore) {
        self.store = store
        self.displayName = store.displayName
        self.channelText = store.channelCodeText
    }

    deinit {
        peerTask?.cancel()
        transportTask?.cancel()
    }

    public var appModeText: String {
        store.environment.mode.rawValue
    }

    public var redactedLastError: String {
        transportState.lastError?.redactedDiagnostic ?? "None"
    }

    public func start() {
        guard peerTask == nil, transportTask == nil else { return }

        peerTask = Task { [environment = store.environment] in
            for await peers in environment.engine.observePeers() {
                if Task.isCancelled { return }
                nearbyPeerCount = peers.count
            }
        }

        transportTask = Task { [environment = store.environment] in
            for await state in environment.engine.observeTransportState() {
                if Task.isCancelled { return }
                transportState = state
            }
        }

        Task {
            permissionSnapshot = await store.environment.permissionSnapshot()
            identity = try? await store.environment.engine.localIdentity()
        }
    }

    public func saveDisplayName() async {
        do {
            try await store.updateDisplayName(displayName)
            errorMessage = nil
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func switchChannel() async {
        do {
            let channel = try ChannelCode(channelText)
            try await store.switchChannel(channel)
            channelText = channel.normalized
            errorMessage = nil
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

public extension String {
    var redactedDiagnostic: String {
        if isEmpty { return "None" }
        let cleaned = replacingOccurrences(of: "\n", with: " ")
        if cleaned.count <= 28 { return cleaned }
        return "\(cleaned.prefix(18))...\(cleaned.suffix(6))"
    }
}

File: android-localwave/app/src/main/java/com/localwave/feature/settings/SettingsViewModel.kt
package com.localwave.feature.settings

import com.localwave.app.AppEnvironment
import com.localwave.app.AppStateHolder
import com.localwave.app.PersistedAppState
import com.localwave.core.diagnostics.RedactedLogger
import com.localwave.core.model.ChannelCode
import com.localwave.core.model.LocalIdentity
import com.localwave.core.model.TransportState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class SettingsUiState(
    val displayName: String = "",
    val channelText: String = "",
    val identity: LocalIdentity? = null,
    val transportState: TransportState = TransportState(),
    val nearbyPeerCount: Int = 0,
    val error: String? = null
)

class SettingsViewModel(
    private val environment: AppEnvironment,
    private val appStateHolder: AppStateHolder,
    persisted: PersistedAppState
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val _state = MutableStateFlow(SettingsUiState(displayName = persisted.displayName, channelText = persisted.channelText))
    val state: StateFlow<SettingsUiState> = _state

    fun start() {
        scope.launch { environment.engine.observeTransportState().collectLatest { t -> _state.update { it.copy(transportState = t) } } }
        scope.launch { environment.engine.observePeers().collectLatest { peers -> _state.update { it.copy(nearbyPeerCount = peers.size) } } }
        scope.launch { _state.update { it.copy(identity = environment.engine.localIdentity()) } }
    }

    fun updateDisplayName(value: String) = _state.update { it.copy(displayName = value) }
    fun updateChannel(value: String) = _state.update { it.copy(channelText = value) }

    fun saveDisplayName() = scope.launch {
        environment.engine.updateDisplayName(state.value.displayName)
        appStateHolder.updateDisplayName(state.value.displayName)
    }

    fun switchChannel() = scope.launch {
        try {
            val channel = ChannelCode(state.value.channelText)
            environment.engine.switchChannel(channel)
            appStateHolder.updateChannel(channel)
            _state.update { it.copy(channelText = channel.normalized, error = null) }
        } catch (error: IllegalArgumentException) {
            _state.update { it.copy(error = error.message) }
        }
    }

    fun redactedLastError(): String = state.value.transportState.lastError?.let { RedactedLogger.redact(it) } ?: "None"
}
